syntax = "proto3";

import "google/protobuf/any.proto";
import "ubiquity/serialization/Shoebox.proto";
import "ubiquity/serialization/Field.proto";
import "ubiquity/serialization/Method.proto";

message WavePB {
  WaveHeaderPB header = 1;
  oneof data {
    ShoeboxPB shoebox = 10;
    FieldGetRequestPB field_get_request = 11;
    FieldGetResponsePB field_get_response = 12;
    FieldSetRequestPB field_set_request = 13;
    FieldSetResponsePB field_set_response = 14;
    MethodCallRequestPB method_call_request = 15;
    MethodCallResponsePB method_call_response = 16;
    ErrorPB error = 17;
  }
}


enum WaveTypePB {
  NONE = 0;
  SHOEBOX = 1;
  FIELD_GETTER_REQUEST = 2;
  FIELD_GETTER_RESPONSE = 3;
  FIELD_SETTER_REQUEST = 4;
  FIELD_SETTER_RESPONSE = 5;
  METHOD_CALL_REQUEST = 6;
  METHOD_CALL_RESPONSE = 7;
  ERROR = 20;
}


message WaveHeaderPB {
  string id = 1;
  WaveTypePB type = 2;
  string shoebox = 3;
  int64 quantum_id = 4;
  string request_wave = 5;
}


message FieldGetRequestPB {
  FieldPB field = 1;
}


message FieldGetResponsePB {
  google.protobuf.Any return_value = 2;
}


message FieldSetRequestPB {
  FieldPB field = 1;
  google.protobuf.Any field_value = 2;
}


message FieldSetResponsePB {}


message MethodCallRequestPB {
  MethodPB method = 1;
  repeated ParameterPB arguments = 2;
}


message MethodCallResponsePB {
  google.protobuf.Any return_value = 2;
}


message ErrorPB {
  string etype = 1;
  string emessage = 2;
  string etrace = 3;
}
